File: java/src/test/java/test_service/models/ModelsTest.java
package test_service.models;

import com.fasterxml.jackson.databind.*;
import org.junit.Assert;
import org.junit.Test;

import java.io.*;
import java.math.BigDecimal;
import java.time.*;
import java.util.*;

public class ModelsTest {

	public static String fixQuotes(String jsonStr) {
		return jsonStr.replaceAll("'", "\"");
	}

	public <T> void check(T data, String jsonStr, Class<T> tClass) throws IOException {
		ObjectMapper objectMapper = new ObjectMapper();

		String actualJson = Jsoner.serialize(objectMapper, data);
		jsonStr = fixQuotes(jsonStr);
		Assert.assertEquals(jsonStr, actualJson);

		T actualData = Jsoner.deserialize(objectMapper, jsonStr, tClass);
		Assert.assertEquals(actualData, data);
	}

	public <T> void checkToString(T data, String expected) {
		String dataStr = data.toString();
		Assert.assertEquals(dataStr, expected);
	}

	@Test
	public void jsonMessageTest() throws IOException {
		Message data = new Message(123);
		String jsonStr = "{'field':123}";
		check(data, jsonStr, Message.class);

		String expected = "Message{field=123}";
		checkToString(data, expected);
	}

	@Test
	public void jsonNestedTest() throws IOException {
		Nested data = new Nested("the string");
		String jsonStr = "{'field':'the string'}";
		check(data, jsonStr, Nested.class);

		String expected = "Nested{field=the string}";
		checkToString(data, expected);
	}

	@Test
	public void jsonParentTest() throws IOException {
		Parent data = new Parent("the string", new Nested("the nested string"));
		String jsonStr = "{'field':'the string','nested':{'field':'the nested string'}}";
		check(data, jsonStr, Parent.class);

		String expected = "Parent{field=the string, nested=Nested{field=the nested string}}";
		checkToString(data, expected);
	}

	@Test
	public void jsonEnumTest() throws IOException {
		EnumFields data = new EnumFields(Choice.SECOND_CHOICE);
		String jsonStr = "{'enum_field':'SECOND_CHOICE'}";
		check(data, jsonStr, EnumFields.class);

		String expected = "EnumFields{enumField=SECOND_CHOICE}";
		checkToString(data, expected);
	}

	@Test
	public void jsonNumericFieldsTest() throws IOException {
		NumericFields data = new NumericFields(123, 1234, 1.23f, 1.23, new BigDecimal("1.23"));
		String jsonStr = "{'int_field':123,'long_field':1234,'float_field':1.23,'double_field':1.23,'decimal_field':1.23}";
		check(data, jsonStr, NumericFields.class);

		String expected = "NumericFields{intField=123, longField=1234, floatField=1.23, doubleField=1.23, decimalField=1.23}";
		checkToString(data, expected);
	}

	@Test
	public void jsonNonNumericFieldsTest() throws IOException {
		NonNumericFields data = new NonNumericFields(true, "the string", UUID.fromString("123e4567-e89b-12d3-a456-426655440000"), LocalDate.parse("2019-11-30"), LocalDateTime.parse("2019-11-30T17:45:55"));
		String jsonStr = "{'boolean_field':true,'string_field':'the string','uuid_field':'123e4567-e89b-12d3-a456-426655440000','date_field':'2019-11-30','datetime_field':'2019-11-30T17:45:55'}";
		check(data, jsonStr, NonNumericFields.class);

		String expected = "NonNumericFields{booleanField=true, stringField=the string, uuidField=123e4567-e89b-12d3-a456-426655440000, dateField=2019-11-30, datetimeField=2019-11-30T17:45:55}";
		checkToString(data, expected);
	}

	@Test
	public void jsonArrayFieldsTest() throws IOException {
		ArrayFields data = new ArrayFields(new int[]{1, 2, 3}, new String[]{"one", "two", "three"});
		String jsonStr = "{'int_array_field':[1,2,3],'string_array_field':['one','two','three']}";
		check(data, jsonStr, ArrayFields.class);

		String expected = "ArrayFields{intArrayField=[1, 2, 3], stringArrayField=[one, two, three]}";
		checkToString(data, expected);
	}

	@Test
	public void jsonMapFieldsTest() throws IOException {
		MapFields data = new MapFields(new HashMap<String, Integer>() {{
			put("one", 1);
			put("two", 2);
		}}, new HashMap<String, String>() {{
			put("one", "first");
			put("two", "second");
		}});
		String jsonStr = "{'int_map_field':{'one':1,'two':2},'string_map_field':{'one':'first','two':'second'}}";
		check(data, jsonStr, MapFields.class);

		String expected = "MapFields{intMapField={one=1, two=2}, stringMapField={one=first, two=second}}";
		checkToString(data, expected);
	}

	@Test
	public void jsonOptionalFieldsTest() throws IOException {
		OptionalFields data = new OptionalFields(123, "the string");
		String jsonStr = "{'int_option_field':123,'string_option_field':'the string'}";
		check(data, jsonStr, OptionalFields.class);

		String expected = "OptionalFields{intOptionField=123, stringOptionField=the string}";
		checkToString(data, expected);
	}

	@Test
	public void jsonRawJsonFieldTest() throws IOException {
		String jsonField = fixQuotes("{'the_array':[1,'some string'],'the_object':{'the_bool':true,'the_string':'some value'},'the_scalar':123}");
		RawJsonField data = new RawJsonField(jsonField);
		String jsonStr = "{'json_field':{'the_array':[1,'some string'],'the_object':{'the_bool':true,'the_string':'some value'},'the_scalar':123}}";
		check(data, jsonStr, RawJsonField.class);

		String expected = fixQuotes("RawJsonField{jsonField={'the_array':[1,'some string'],'the_object':{'the_bool':true,'the_string':'some value'},'the_scalar':123}}");
		checkToString(data, expected);
	}

	@Test
	public void jsonOrderCreatedTest() throws IOException {
		OrderCreated data = new OrderCreated(UUID.fromString("58d5e212-165b-4ca0-909b-c86b9cee0111"), "SNI/01/136/0500", 3);
		String jsonStr = "{'id':'58d5e212-165b-4ca0-909b-c86b9cee0111','sku':'SNI/01/136/0500','quantity':3}";
		check(data, jsonStr, OrderCreated.class);

		String expected = "OrderCreated{id=58d5e212-165b-4ca0-909b-c86b9cee0111, sku=SNI/01/136/0500, quantity=3}";
		checkToString(data, expected);
	}

	@Test
	public void jsonOneOfWrapperTest() throws IOException {
		OrderEvent data = new OrderEventCanceled(new OrderCanceled(UUID.fromString("123e4567-e89b-12d3-a456-426655440000")));
		String jsonStr = "{'canceled':{'id':'123e4567-e89b-12d3-a456-426655440000'}}";
		check(data, jsonStr, OrderEvent.class);

		String expected = "OrderCanceled{data=OrderCanceled{id=123e4567-e89b-12d3-a456-426655440000}}";
		checkToString(data, expected);
	}

	@Test
	public void jsonOneOfDiscriminatorTest() throws IOException {
		OrderEventDiscriminated data = new OrderEventDiscriminatedCanceled(new OrderCanceled(UUID.fromString("123e4567-e89b-12d3-a456-426655440000")));
		String jsonStr = "{'_type':'canceled','id':'123e4567-e89b-12d3-a456-426655440000'}";
		check(data, jsonStr, OrderEventDiscriminated.class);

		String expected = "OrderCanceled{data=OrderCanceled{id=123e4567-e89b-12d3-a456-426655440000}}";
		checkToString(data, expected);
	}
}

File: go/tests/models_test.go
package tests

import (
	"cloud.google.com/go/civil"
	"encoding/json"
	"github.com/google/uuid"
	"github.com/shopspring/decimal"
	"gotest.tools/v3/assert"
	"reflect"
	"test-models/spec/models"
	"testing"
)

func TestMessageFields(t *testing.T) {
	data := models.Message{
		0,
	}

	jsonStr := `{"field":0}`

	actualJson, err := json.Marshal(data)
	assert.NilError(t, err)
	assert.Equal(t, jsonStr, string(actualJson))

	var actualData models.Message
	err = json.Unmarshal([]byte(jsonStr), &actualData)
	assert.NilError(t, err)
	assert.Equal(t, reflect.DeepEqual(data, actualData), true)
}

func TestNestedFields(t *testing.T) {
	data := models.Parent{
		"the string",
		models.Nested{
			"the nested string",
		},
	}

	jsonStr := `{"field":"the string","nested":{"field":"the nested string"}}`

	actualJson, err := json.Marshal(data)
	assert.NilError(t, err)
	assert.Equal(t, jsonStr, string(actualJson))

	var actualData models.Parent
	err = json.Unmarshal([]byte(jsonStr), &actualData)
	assert.NilError(t, err)
	assert.Equal(t, reflect.DeepEqual(data, actualData), true)
}

func TestNumericFields(t *testing.T) {
	decimalField, _ := decimal.NewFromString("1.23")
	data := models.NumericFields{
		0,
		0,
		1.23,
		1.23,
		decimalField,
	}

	jsonStr := `{"int_field":0,"long_field":0,"float_field":1.23,"double_field":1.23,"decimal_field":"1.23"}`

	actualJson, err := json.Marshal(data)
	assert.NilError(t, err)
	assert.Equal(t, jsonStr, string(actualJson))

	var actualData models.NumericFields
	err = json.Unmarshal([]byte(jsonStr), &actualData)
	assert.NilError(t, err)
	assert.Equal(t, reflect.DeepEqual(data, actualData), true)
}

func TestNumericFieldsNegative(t *testing.T) {
	jsonStr := `{"int_field":0,"long_field":0,"float_field":1.23,"double_field":1.23,"decimal_field":nil}`

	var actualData models.NumericFields
	err := json.Unmarshal([]byte(jsonStr), &actualData)
	assert.ErrorContains(t, err, "null")
}

func TestNonNumericFields(t *testing.T) {
	dateField, _ := civil.ParseDate("2019-11-30")
	timeDateField, _ := civil.ParseDateTime("2019-11-30T17:45:55")
	data := models.NonNumericFields{
		true,
		"the string",
		uuid.MustParse("123e4567-e89b-12d3-a456-426655440000"),
		dateField,
		timeDateField,
	}

	jsonStr := `{"boolean_field":true,"string_field":"the string","uuid_field":"123e4567-e89b-12d3-a456-426655440000","date_field":"2019-11-30","datetime_field":"2019-11-30T17:45:55"}`

	actualJson, err := json.Marshal(data)
	assert.NilError(t, err)
	assert.Equal(t, jsonStr, string(actualJson))

	var actualData models.NonNumericFields
	err = json.Unmarshal([]byte(jsonStr), &actualData)
	assert.NilError(t, err)
	assert.Equal(t, reflect.DeepEqual(data, actualData), true)
}

func TestNonNumericFieldsNegative(t *testing.T) {
	jsonStr := `{"boolean_field":true,"string_field":"the string","uuid_field":nil,"date_field":"2019-11-30","datetime_field":"2019-11-30T17:45:55"}`

	var actualData models.NonNumericFields
	err := json.Unmarshal([]byte(jsonStr), &actualData)
	assert.ErrorContains(t, err, "null")
}

func TestArrayFields(t *testing.T) {
	data := models.ArrayFields{
		[]int{1, 2, 3},
		[]string{"one", "two", "three"},
	}

	jsonStr := `{"int_array_field":[1,2,3],"string_array_field":["one","two","three"]}`

	actualJson, err := json.Marshal(data)
	assert.NilError(t, err)
	assert.Equal(t, jsonStr, string(actualJson))

	var actualData models.ArrayFields
	err = json.Unmarshal([]byte(jsonStr), &actualData)
	assert.NilError(t, err)
	assert.Equal(t, reflect.DeepEqual(data, actualData), true)
}

func TestArrayFieldsNegative(t *testing.T) {
	jsonStr := `{"int_array_field":[nil],"string_array_field":["one","two","three"]}`

	var actualData models.ArrayFields
	err := json.Unmarshal([]byte(jsonStr), &actualData)
	assert.ErrorContains(t, err, "null")
}

func TestMapFields(t *testing.T) {
	data := models.MapFields{
		map[string]int{"one": 1, "two": 2},
		map[string]string{"one": "first", "two": "second"},
	}

	jsonStr := `{"int_map_field":{"one":1,"two":2},"string_map_field":{"one":"first","two":"second"}}`

	actualJson, err := json.Marshal(data)
	assert.NilError(t, err)
	assert.Equal(t, jsonStr, string(actualJson))

	var actualData models.MapFields
	err = json.Unmarshal([]byte(jsonStr), &actualData)
	assert.NilError(t, err)
	assert.Equal(t, reflect.DeepEqual(data, actualData), true)
}

func TestMapFieldsNegative(t *testing.T) {
	jsonStr := `{"int_map_field":{"one":1,"two":2},"string_map_field":{"one":nil,"two":"second"}}`

	var actualData models.MapFields
	err := json.Unmarshal([]byte(jsonStr), &actualData)
	assert.ErrorContains(t, err, "null")
}

func TestOptionalFields(t *testing.T) {
	var optionalInt = 123
	var optionalString = "the string"
	data := models.OptionalFields{
		&optionalInt,
		&optionalString,
	}

	jsonStr := `{"int_option_field":123,"string_option_field":"the string"}`

	actualJson, err := json.Marshal(data)
	assert.NilError(t, err)
	assert.Equal(t, jsonStr, string(actualJson))

	var actualData models.OptionalFields
	err = json.Unmarshal([]byte(jsonStr), &actualData)
	assert.NilError(t, err)
	assert.Equal(t, reflect.DeepEqual(data, actualData), true)
}

func TestOptionalNil(t *testing.T) {
	data := models.OptionalFields{
		nil,
		nil,
	}

	jsonStr := `{"int_option_field":null,"string_option_field":null}`

	actualJson, err := json.Marshal(data)
	assert.NilError(t, err)
	assert.Equal(t, jsonStr, string(actualJson))

	var actualData models.OptionalFields
	err = json.Unmarshal([]byte(jsonStr), &actualData)
	assert.NilError(t, err)
	assert.Equal(t, reflect.DeepEqual(data, actualData), true)
}

func TestRawJsonField(t *testing.T) {
	data := models.RawJsonField{
		json.RawMessage(`{"the_array":[1,"some string"],"the_object":{"the_bool":true,"the_string":"some value"},"the_scalar":123}`),
	}

	jsonStr := `{"json_field":{"the_array":[1,"some string"],"the_object":{"the_bool":true,"the_string":"some value"},"the_scalar":123}}`

	actualJson, err := json.Marshal(data)
	assert.NilError(t, err)
	assert.Equal(t, jsonStr, string(actualJson))

	var actualData models.RawJsonField
	err = json.Unmarshal([]byte(jsonStr), &actualData)
	assert.NilError(t, err)
	assert.Equal(t, reflect.DeepEqual(data, actualData), true)
}

func TestRawJsonFieldNegative(t *testing.T) {
	jsonStr := `{"json_field":{"the_array":[1,"some string"],"the_object":{"the_bool":nil,"the_string":"some value"},"the_scalar":123}}`

	var actualData models.RawJsonField
	err := json.Unmarshal([]byte(jsonStr), &actualData)
	assert.ErrorContains(t, err, "null")
}

func TestOrderCreated(t *testing.T) {
	data := models.OrderCreated{
		uuid.MustParse("58d5e212-165b-4ca0-909b-c86b9cee0111"),
		"SNI/01/136/0500",
		3,
	}

	jsonStr := `{"id":"58d5e212-165b-4ca0-909b-c86b9cee0111","sku":"SNI/01/136/0500","quantity":3}`

	actualJson, err := json.Marshal(data)
	assert.NilError(t, err)
	assert.Equal(t, jsonStr, string(actualJson))

	var actualData models.OrderCreated
	err = json.Unmarshal([]byte(jsonStr), &actualData)
	assert.NilError(t, err)
	assert.Equal(t, reflect.DeepEqual(data, actualData), true)
}

func TestOrderCreatedNegative(t *testing.T) {
	jsonStr := `{"id":"58d5e212-165b-4ca0-909b-c86b9cee0111","sku":"SNI/01/136/0500","quantity":nil}`

	var actualData models.OrderCreated
	err := json.Unmarshal([]byte(jsonStr), &actualData)
	assert.ErrorContains(t, err, "null")
}

func TestEnumFields(t *testing.T) {
	data := models.EnumFields{
		models.ChoiceSecondChoice,
	}

	jsonStr := `{"enum_field":"SECOND_CHOICE"}`

	actualJson, err := json.Marshal(data)
	assert.NilError(t, err)
	assert.Equal(t, jsonStr, string(actualJson))

	var actualData models.EnumFields
	err = json.Unmarshal([]byte(jsonStr), &actualData)
	assert.NilError(t, err)
	assert.Equal(t, reflect.DeepEqual(data, actualData), true)
}

func TestEnumFieldsNegative(t *testing.T) {
	jsonStr := `{"enum_field":"gfgnfg"}`

	var actualData models.EnumFields
	err := json.Unmarshal([]byte(jsonStr), &actualData)
	assert.ErrorContains(t, err, "gfgnfg")
}

func TestOneOfFields(t *testing.T) {
	data := models.OrderEvent{
		Changed: &models.OrderChanged{uuid.MustParse("58d5e212-165b-4ca0-909b-c86b9cee0111"), 3},
	}

	jsonStr := `{"changed":{"id":"58d5e212-165b-4ca0-909b-c86b9cee0111","quantity":3}}`

	actualJson, err := json.Marshal(data)
	assert.NilError(t, err)
	assert.Equal(t, jsonStr, string(actualJson))

	var actualData models.OrderEvent
	err = json.Unmarshal([]byte(jsonStr), &actualData)
	assert.NilError(t, err)
	assert.Equal(t, reflect.DeepEqual(data, actualData), true)
}

func TestDiscriminatedOneOfFields(t *testing.T) {
	data := models.OrderEventDiscriminated{
		Changed: &models.OrderChanged{uuid.MustParse("58d5e212-165b-4ca0-909b-c86b9cee0111"), 3},
	}

	jsonStr := `{"_type":"changed","id":"58d5e212-165b-4ca0-909b-c86b9cee0111","quantity":3}`

	actualJson, err := json.Marshal(data)
	assert.NilError(t, err)
	assert.Equal(t, jsonStr, string(actualJson))

	var actualData models.OrderEventDiscriminated
	err = json.Unmarshal([]byte(jsonStr), &actualData)
	assert.NilError(t, err)
	assert.Equal(t, reflect.DeepEqual(data, actualData), true)
}
